File: Problems/11/main.go
package main

func maxArea(height []int) int {
	res := 0
	i, j := 0, len(height)-1

	for i < j {
		cand := (j - i) * min(height[i], height[j])

		if cand > res {
			res = cand
		}

		if height[i] < height[j] {
			i += 1
		} else if height[i] > height[j] {
			j -= 1
		} else {
			i += 1
			j -= 1
		}
	}

	return int(res)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: Problems/11/main.cpp
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>

using namespace std;

class Solution
{
public:
    int maxArea(vector<int> &H)
    {
        int l = 0, r = H.size() - 1;

        int res = INT_MIN;
        while (l < r)
        {
            res = max(res, (r - l) * min(H[l], H[r]));
            if (H[l] < H[r])
                l++;
            else
                r--;
        }
        return res;
    }
};
